File: day1/day1.cpp
#include <fstream>
#include <vector>

#include <fmt/format.h>

using namespace std;

static size_t solve(const vector<int>& depths, size_t l)
{
	size_t count = 0;
	if (depths.size() >= l)
	{
		int sum = 0;
		for (size_t i = 0; i < l; i++)
		{
			sum += depths[i];
		}
		for (size_t i = l; i < depths.size(); i++)
		{
			int nsum = sum + depths[i] - depths[i-l];
			if (nsum > sum)
			{
				count++;
			}
			sum = nsum;
		}
	}
	return count;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}", argv[1]);
		return 1;
	}

	vector<int> depths;
	int depth;
	while (in >> depth)
	{
		depths.push_back(depth);
	}
	in.close();

	fmt::print("Part1: {}\n", solve(depths, 1));
	fmt::print("Part2: {}\n", solve(depths, 3));

	return 0;
}

File: day1/day1.py
#!/usr/bin/env python3

import sys

def part1(depths):
    count = 0
    prev = depths[0]
    for cur in depths[1:]:
        if cur > prev:
            count += 1
        prev = cur
    return count

def solve(depths, l):
    if len(depths) < l:
        return 0

    count = 0
    csum = sum(depths[:l])
    for i in range(l, len(depths)):
        nsum = csum + depths[i] - depths[i-l]
        if csum < nsum:
            count += 1
        csum = nsum

    return count

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            depths = [int(x) for x in f.read().strip().split()]

        print("Part1:", solve(depths, 1))
        print("Part2:", solve(depths, 3))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)
